bouncer/shared/executable_test.ts

```diff
@@ -0,0 +1,94 @@
+import { SwapContext } from './swap_context';
+import { runWithTimeout } from './utils';
+
+export enum TestStatus {
+  Ready = 'ready',
+  Running = 'running',
+  Complete = 'complete',
+  Failed = 'failed',
+}
+
+export class ExecutableTest {
+  public status: TestStatus = TestStatus.Ready;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private runFunction: (...args: any[]) => Promise<void>;
+
+  timeoutSeconds: number;
+
+  debug = false;
+
+  public swapContext: SwapContext;
+
+  constructor(
+    public name: string,
+    testFunction: () => Promise<void>,
+    timeoutSeconds: number,
+  ) {
+    this.runFunction = testFunction;
+    this.timeoutSeconds = timeoutSeconds;
+    this.swapContext = new SwapContext();
+  }
+
+  /// Run the test with the pre-defined timeout and any given arguments
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async run(...args: any[]) {
+    if (this.status === TestStatus.Running) {
+      throw new Error(`${this.name} Test is already running`);
+    }
+    console.log('\x1b[36m%s\x1b[0m', `=== Running ${this.name} test ===`);
+    this.status = TestStatus.Running;
+    await runWithTimeout(this.runFunction(...args), this.timeoutSeconds * 1000).catch((error) => {
+      // Print the swap report if the swap context was used
+      if (this.swapContext.allSwaps.size > 0) {
+        this.swapContext.print_report();
+      }
```

Comment:
This seems out of place to me - it seems like something that should be the responsibility of the swapping test - not of the general executable test function. Here it would probably look like something that the test being called can provide back to the caller like "callOnTimeout()" or "callOnFailure()", which would be `this.swapContext.print_report()` for the swapping test for example

Reply:
I disagree. Having it auto print the report is better than having to remember to add code to every test that will use the swap context to output the report on failure.
The `callOnTimeout` and `callOnFailure` is a good idea for when a test has custom stuff it needs to do, like cleaning code, but so far no tests need this, so I will add it when required.

Reply:
Not every test is a swapping test is my point though, it's not relevant for every test, and therefore shouldn't be here

Reply:
The swap context is a property of the executable test, so it seems fine to make printing the report it's responsibility.  
I agree that it's not perfect, but doing it another way will add complexity for no clear benefit.